src/implicit.rs

```diff
@@ -256,6 +256,7 @@ pub fn resolve_neutral<'a, D: BidiDataSource, T: TextSource<'a> + ?Sized>(
     original_classes: &[BidiClass],
     processing_classes: &mut [BidiClass],
 ) where
+    <T as TextSource<'a>>::CharIndexIter: Iterator<Item = (usize, char)>,
```

Comment:
nit: all of these bounds should live on the TextSource trait, not on all these where clauses